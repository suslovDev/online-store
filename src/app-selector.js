const state = {
  products: {
    products: [],
    originProducts: [
      {
        id: 1,
        barcode: 12345,
        brand: "Brand",
        description: "description",
        manufacturer: "Manufacturer",
        price: 199,
        size: 367,
        sizeType: "weight",
        title: "Title",
        url: "/url",
        care: [],
      },
    ],
    order: {
      type: "name",
      order: "increase",
    },
    filter: {
      priceRange: { from: 0, to: 10000 },
      substr: "",
      care: "",
      manufacturer: [],
    },
  },
  cart: {
    cart: [
      {
        id: 1,
        barcode: 12345678910,
        brand: "Brand",
        description: "description",
        manufacturer: "Manufacturer",
        price: 199,
        size: 250,
        sizeType: "weight",
        title: "Title",
        url: "./test-url",
        care: ["care1", "care2"],
      },
    ],
    total: 0,
  },
};

export const AppSelector = (f) => f(state);
