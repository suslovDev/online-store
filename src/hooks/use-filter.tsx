import { useEffect } from "react";
import { useAppDispatch, useAppSelector } from "./hooks";
import { filterProducts, setFilter } from "../store/products-slice";

export const useFilter = () => {
  const filter = useAppSelector((state) => state.products.filter);

  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(filterProducts(filter));
  }, [filter, dispatch]);

  const filterProd = (p: any) => { //не забыть описать типы!
    let finalFilter = { ...filter, ...p };
    dispatch(setFilter(finalFilter));
  };
  return filterProd;
};
