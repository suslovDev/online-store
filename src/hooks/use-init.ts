import { useState, useEffect } from "react";
import { useAppSelector } from "./hooks";

export const useInit = () => {
    const { originProducts } = useAppSelector((state) => state.products);
    const [init, setInit] = useState(false);
    if (originProducts.length && init) localStorage.removeItem("products");

    useEffect(() => {
        if (init) return;
        setInit(true);
    }, [originProducts]);

    return originProducts;
}