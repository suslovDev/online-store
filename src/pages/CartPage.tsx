import { FC, useState } from "react";
import classes from "./CartPage.module.css";
import { clearCart } from "../store/cart-slice";
import { useAppSelector, useAppDispatch } from "../hooks/hooks";
import Button from "../components/UI/Button";
import CartItem from "../components/Cart/cart-item/CartItem";
import Container from "../components/Layout/Container/Container";
import Crumbs from "../components/Layout/Crumbs/Crumbs";
import HorisontalGap from "../components/Layout/HorisontalGap/HorisontalGap";
import ThanksOrderModal from "../components/Cart/modal/ThanksOrderModal";
import Price from "../components/UI/Price";

const params = [
  { url: "", name: "Главная" },
  { url: "cart", name: "Корзина" },
];

export const CartPage: FC = () => {
  const { cart, total } = useAppSelector((state) => state.cart);
  const [isOpen, setIsopen] = useState(false);

  const dispatch = useAppDispatch();

  const totalValue = total.toFixed(2);

  const handleOrderDone = () => {
    setIsopen(false);
    dispatch(clearCart());
  };

  return (
    <Container>
      {isOpen && <ThanksOrderModal onClick={handleOrderDone} />}
      <Crumbs params={params} />
      <h2 className={classes.heading}>{`Корзина ${
        cart.length ? "" : "пуста!"
      }`}</h2>
      <HorisontalGap gap='50px' />
      <ul>
        {cart.map((item) => (
          <li>
            <CartItem item={item} />
          </li>
        ))}
      </ul>
      {cart.length !== 0 && (
        <div className={classes.checkout}>
          <Button variant='standart' onClick={() => setIsopen(true)}>
            Оформить заказ
          </Button>
          <Price currency='KZT' type='large' value={+totalValue} />
        </div>
      )}
    </Container>
  );
};
