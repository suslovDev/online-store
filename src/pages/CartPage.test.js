import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { useAppSelector } from "../hooks/hooks";
import { AppSelector } from "../app-selector";
import { CartPage } from "./CartPage";

jest.mock("../hooks/hooks");

const mockAppSelector = useAppSelector;
mockAppSelector.mockImplementation(AppSelector);

describe("CartPage.tsx:", () => {
  beforeEach(() => {
    mockAppSelector.mockImplementation(AppSelector);
  });
  it("корзина наполняется товарами", () => {
    const component = render(
      <MemoryRouter>
        <CartPage />
      </MemoryRouter>
    );
    expect(component).toMatchSnapshot();
  });
});
