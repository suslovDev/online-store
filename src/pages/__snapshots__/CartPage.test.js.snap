// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartPage.tsx: корзина наполняется товарами 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <div
          style="margin-bottom: 45px;"
        />
        <nav>
          <ul
            class="crumbs"
          >
            <li>
              <a
                href="/"
              >
                Главная
              </a>
            </li>
            <li>
              <a
                href="/cart"
              >
                Корзина
              </a>
            </li>
          </ul>
        </nav>
        <div
          style="margin-bottom: 65px;"
        />
        <h2
          class="heading"
          data-testid="cart-page"
        >
          Корзина 
        </h2>
        <div
          style="margin-bottom: 50px;"
        />
        <ul>
          <li>
            <div
              class="item"
            >
              <div
                class="image"
              >
                <img
                  alt=""
                  src="./test-url"
                />
              </div>
              <div
                class="content"
              >
                <div
                  class="info"
                >
                  <div
                    class="wrap"
                  >
                    <img
                      alt=""
                      class="img"
                      src="/weight-img.png"
                    />
                    <span
                      class="value"
                    >
                      250 гр.
                    </span>
                  </div>
                  <p
                    class="title"
                  >
                    Title description
                  </p>
                  <p
                    class="description"
                  >
                    description
                  </p>
                </div>
                <div
                  class="actions"
                >
                  <div
                    class="wrap"
                  >
                    <div
                      class="amount"
                    >
                      <button>
                        -
                      </button>
                      <span />
                      <button>
                        +
                      </button>
                    </div>
                  </div>
                  <div
                    class="wrap"
                  >
                    <span
                      class="large"
                      data-testid="price"
                    >
                      199 ₸
                    </span>
                  </div>
                  <div
                    class="wrap"
                    data-testid="remove"
                  >
                    <button
                      class="button round-large"
                    >
                      <img
                        alt=""
                        src="./remove-img.png"
                        width="27"
                      />
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </li>
        </ul>
        <div
          class="checkout"
        >
          <button
            class="button large"
          >
            Оформить заказ
          </button>
          <span
            class="large"
            data-testid="price"
          >
            0 ₸
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div
        style="margin-bottom: 45px;"
      />
      <nav>
        <ul
          class="crumbs"
        >
          <li>
            <a
              href="/"
            >
              Главная
            </a>
          </li>
          <li>
            <a
              href="/cart"
            >
              Корзина
            </a>
          </li>
        </ul>
      </nav>
      <div
        style="margin-bottom: 65px;"
      />
      <h2
        class="heading"
        data-testid="cart-page"
      >
        Корзина 
      </h2>
      <div
        style="margin-bottom: 50px;"
      />
      <ul>
        <li>
          <div
            class="item"
          >
            <div
              class="image"
            >
              <img
                alt=""
                src="./test-url"
              />
            </div>
            <div
              class="content"
            >
              <div
                class="info"
              >
                <div
                  class="wrap"
                >
                  <img
                    alt=""
                    class="img"
                    src="/weight-img.png"
                  />
                  <span
                    class="value"
                  >
                    250 гр.
                  </span>
                </div>
                <p
                  class="title"
                >
                  Title description
                </p>
                <p
                  class="description"
                >
                  description
                </p>
              </div>
              <div
                class="actions"
              >
                <div
                  class="wrap"
                >
                  <div
                    class="amount"
                  >
                    <button>
                      -
                    </button>
                    <span />
                    <button>
                      +
                    </button>
                  </div>
                </div>
                <div
                  class="wrap"
                >
                  <span
                    class="large"
                    data-testid="price"
                  >
                    199 ₸
                  </span>
                </div>
                <div
                  class="wrap"
                  data-testid="remove"
                >
                  <button
                    class="button round-large"
                  >
                    <img
                      alt=""
                      src="./remove-img.png"
                      width="27"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </li>
      </ul>
      <div
        class="checkout"
      >
        <button
          class="button large"
        >
          Оформить заказ
        </button>
        <span
          class="large"
          data-testid="price"
        >
          0 ₸
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
